File: 53.maximum-subarray.go
// @leet start
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxSubArray(nums []int) int {
	currentMax := nums[0]
	totalMax := nums[0]
	for i := 1; i<len(nums); i++ {
		currentMax = max(nums[i], currentMax+nums[i])
		totalMax = max(totalMax, currentMax)
	}
	return totalMax
}
// @leet end

File: 53.maximum-subarray.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number}
 */
// Space Complexity = O(1)
// Time Complexity = O(n)
var maxSubArray = function(nums) {
	let maxTotal = nums[0];
	let currentTotal = 0;

	for(let num of nums) {
		currentTotal = Math.max(num, currentTotal + num)
		maxTotal = Math.max(currentTotal, maxTotal)
	}
	return maxTotal
};
// @leet end
